asserts/preseed.go

```diff
@@ -0,0 +1,218 @@
+func (p *Preseed) BrandID() string {
+	return p.HeaderString("brand-id")
+}
+
+// Model returns the model name identifier.
+func (p *Preseed) Model() string {
+	return p.HeaderString("model")
+}
+
+// SystemLabel returns the label of the seeded system.
+func (p *Preseed) SystemLabel() string {
+	return p.HeaderString("system-label")
+}
+
+// Timestamp returns the time when the preseed assertion was issued.
+func (p *Preseed) Timestamp() time.Time {
+	return p.timestamp
+}
+
+// PreseedSHA3_384 returns the checksum of preseeding artifact.
+func (p *Preseed) PreseedSHA3_384() string {
+	return p.HeaderString("preseed-sha3-384")
+}
+
+// Snaps returns the snaps for preseeding
+func (p *Preseed) Snaps() []*PreseedSnap {
+	return p.snaps
+}
+
+func checkPreseedSnap(snap map[string]interface{}) (*PreseedSnap, error) {
+	name, err := checkNotEmptyStringWhat(snap, "name", "of snap")
+	if err != nil {
+		return nil, err
+	}
+	if err := naming.ValidateSnap(name); err != nil {
+		return nil, fmt.Errorf("invalid snap name %q", name)
+	}
+
+	what := fmt.Sprintf("of snap %q", name)
+
+	var snapID string
+	if _, ok := snap["id"]; ok {
+		snapID, err = checkStringMatchesWhat(snap, "id", what, naming.ValidSnapID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var snapRevision int
+	if _, ok := snap["revision"]; ok {
+		var err error
+		snapRevision, err = checkSnapRevisionWhat(snap, "revision", what)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if snapID != "" && snapRevision <= 0 {
+		return nil, fmt.Errorf("snap revision is required when snap id is set")
+	}
+	if snapID == "" && snapRevision > 0 {
+		return nil, fmt.Errorf("snap id is required when revision is set")
+	}
+
+	return &PreseedSnap{
+		Name:     name,
+		SnapID:   snapID,
+		Revision: snapRevision,
+	}, nil
+}
+
+func checkPreseedSnaps(snapList interface{}) ([]*PreseedSnap, error) {
+	const wrongHeaderType = `"snaps" header must be a list of maps`
+
+	entries, ok := snapList.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf(wrongHeaderType)
+	}
+
+	seen := make(map[string]bool, len(entries))
+	seenIDs := make(map[string]string, len(entries))
+	snaps := make([]*PreseedSnap, 0, len(entries))
+	for _, entry := range entries {
+		snap, ok := entry.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf(wrongHeaderType)
+		}
+		preseedSnap, err := checkPreseedSnap(snap)
+		if err != nil {
+			return nil, err
+		}
+
+		if seen[preseedSnap.Name] {
+			return nil, fmt.Errorf("cannot list the same snap %q multiple times", preseedSnap.Name)
+		}
+		seen[preseedSnap.Name] = true
+		snapID := preseedSnap.SnapID
+		if underName := seenIDs[snapID]; underName != "" {
+			return nil, fmt.Errorf("cannot specify the same snap id %q multiple times, specified for snaps %q and %q", snapID, underName, preseedSnap.Name)
+		}
+		seenIDs[snapID] = preseedSnap.Name
+		snaps = append(snaps, preseedSnap)
+	}
+
+	return snaps, nil
+}
+
+var validSystemLabel = regexp.MustCompile("^[0-9]{8}")
```

Comment:
seed/internal internal has already a ValidateUC20SeedSystemLabel  and regexp , we should probably move them here instead and rename the function to follow this package convention is IsValidSystemLabel